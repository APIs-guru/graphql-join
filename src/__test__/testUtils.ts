declare var expect;

import {
  Source,
  DocumentNode,
  GraphQLError,
  GraphQLSchema,
  GraphQLFieldResolver,
  GraphQLResolveInfo,
  ExecutionResult,

  graphql,
  parse,
  print,
  buildSchema,
  printSchema,
  formatError,
} from 'graphql';
import * as _ from 'lodash';

import { joinSchemas, ProxyContext } from '../index';
import { stubSchema } from '../utils';

expect.addSnapshotSerializer({
  print(val) {
    return typeof val === 'string' ?
      val.trim() :
      JSON.stringify(val, null, 2);
  },
  test(val) {
    return true;
  },
});

function fakeFieldResolver(schemaName: string): GraphQLFieldResolver<any,any> {
  return (
    _1, _2, _3,
    {fieldName, parentType}: GraphQLResolveInfo
  ) => {
    return `${schemaName}::${parentType.name}::${fieldName}`;
  }
}

function makeProxy(schemaName: string, schema: GraphQLSchema) {
  return (queryAST: DocumentNode) => {
    const query = print(queryAST);
    expect(query).toMatchSnapshot();
    return graphql({
      schema,
      source: query,
      fieldResolver: fakeFieldResolver(schemaName),
    });
  };
}

type TestSchema = string;
type TestSchemasMap = { [name: string]: TestSchema };
export function testJoin(testSchemas: TestSchemasMap, joinSDL: string) {
  const remoteSchemas = _.mapValues(testSchemas, (sdl, name) => {
    const schema = buildSchema(new Source(sdl, name));
    stubSchema(schema);
    return { schema };
  });

  const joinAST = parse(new Source(joinSDL, 'Join SDL'));
  const schema = joinSchemas(joinAST, remoteSchemas);

  expect(schema).toBeInstanceOf(GraphQLSchema);
  expect(printSchema(schema)).toMatchSnapshot();
  return async (
    query: string,
    results?: { [schemaName: string]: ExecutionResult }
  ) => {
    const proxyFns = _.mapValues(remoteSchemas, ({schema}, name) => {
      if (results && results[name]) {
        return (() => Promise.resolve(results[name]));
      }
      return makeProxy(name, schema)
    });

    const result = await graphql({
      schema,
      source: new Source(query, 'ClientQuery'),
      contextValue: new ProxyContext(proxyFns),
    });
    expect([
      query,
      resultToJSON(result),
    ]).toMatchSnapshot();
  };
}

function resultToJSON(result) {
  return {
    ...result,
    errors: (result.errors || []).map(formatError),
  };
}

