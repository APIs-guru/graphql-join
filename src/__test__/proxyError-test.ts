declare var describe, test;

import { GraphQLError } from 'graphql';
import { testJoin } from './testUtils';

describe('proxy errors', () => {
  const execute = testJoin({
    test: `
      type Query { foo: Bar }
      type Bar { bar: String }
    `,
  },'schema { query: Query }');

  test('global error', async () => {
    await execute('{ foo { bar } }', { test: {
      errors: [ new GraphQLError('global error') ],
    }});
  });
  test('multiple global errors', async () => {
    await execute('{ foo { bar } }', { test: {
      errors: [
        new GraphQLError('first global error'),
        new GraphQLError('second global error')
      ],
    }});
  });
  test('error with path', async () => {
    await execute('{ foo { bar } }', {
      test: {
        data: { foo: { bar: null } },
        errors: [
          new GraphQLError(
            'error with path',
            undefined, undefined, undefined,
            ['foo', 'bar']
          )
        ],
      }
    });
  });

});
