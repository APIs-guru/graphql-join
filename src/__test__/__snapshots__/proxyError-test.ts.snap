// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
type Bar {
  bar: String
}

type Query {
  foo: Bar
}
`;

exports[`proxy errors error with path 1`] = `
[
  "{ foo { bar } }",
  {
    "errors": [
      {
        "message": "error with path",
        "locations": [
          {
            "line": 1,
            "column": 9
          }
        ],
        "path": [
          "foo",
          "bar"
        ]
      }
    ],
    "data": {
      "foo": {
        "bar": null
      }
    }
  }
]
`;

exports[`proxy errors global error 1`] = `
[
  "{ foo { bar } }",
  {
    "errors": [
      {
        "message": "global error",
        "locations": [
          {
            "line": 1,
            "column": 3
          }
        ],
        "path": [
          "foo"
        ]
      }
    ],
    "data": {
      "foo": null
    }
  }
]
`;

exports[`proxy errors multiple global errors 1`] = `
[
  "{ foo { bar } }",
  {
    "errors": [
      {
        "message": "first global error\\nsecond global error",
        "locations": [
          {
            "line": 1,
            "column": 3
          }
        ],
        "path": [
          "foo"
        ]
      }
    ],
    "data": {
      "foo": null
    }
  }
]
`;
